Server/Services/ILocalTempFileLocks.cs

```diff
@@ -14,15 +16,44 @@ public interface ILocalTempFileLocks
     ///   to avoid multiple places using the same temporary folder at once
     /// </summary>
     /// <param name="suffix">Suffix to add to the temporary folder</param>
-    /// <param name="path">The final result path</param>
-    /// <returns>A semaphore that needs to be locked while using the path</returns>
-    public SemaphoreSlim GetTempFilePath(string suffix, out string path);
+    /// <returns>The final result path</returns>
+    public string GetTempFilePath(string suffix);
+
+    /// <summary>
+    /// Asynchronously lock based on a key.
+    /// </summary>
+    /// <param name="path">The file path</param>
+    /// <returns>A disposable value.</returns>
+    public ValueTask<IDisposable> LockAsync(string path);
+
+    /// <summary>
+    /// Asynchronously lock based on a key, while observing a <see cref="CancellationToken"/>.
+    /// </summary>
+    /// <param name="path">The file path</param>
+    /// <param name="cancellationToken">A cancellation token</param>
+    /// <returns>A disposable value.</returns>
+    public ValueTask<IDisposable> LockAsync(string path, CancellationToken cancellationToken);
+
+    /// <summary>
+    /// Asynchronously lock based on a key, setting a limit for the <see cref="TimeSpan"/> to wait,
+    /// while observing a <see cref="CancellationToken"/>.
+    /// </summary>
+    /// <param name="path">The file path</param>
+    /// <param name="timeout">The time limit to wait for</param>
+    /// <param name="cancellationToken">A cancellation token</param>
+    /// <returns>A disposable value.</returns>
+    public ValueTask<AsyncKeyedLockTimeoutReleaser<string>> LockAsync(
+        string path, TimeSpan timeout, CancellationToken cancellationToken);
 }
 
 public class LocalTempFileLocks : ILocalTempFileLocks
 {
     private readonly string baseTempFilePath;
-    private readonly Dictionary<string, SemaphoreSlim> requestedPaths = new();
+    private readonly AsyncKeyedLocker<string> asyncKeyedLocker = new(o =>
+    {
+        o.PoolSize = 20;
+        o.PoolInitialFill = 1;
```

Comment:
Could you add an explanation why these values were selected? I assume 20 doesn't accidentally mean that each string can be locked 20 times (like I'd expect a semaphore to have both initial and max values set to 1).

Reply:
That's the pool of semaphoreslims. It means that as long as there are 1-20 concurrent threads, it would be using everything from the pool. As soon as there's a 21st, it won't wait for one of those 20 to finish but create a new `SemaphoreSlim` anyway. Once things 'calm down' a bit, you'd still be left with 20 items in the pool.

If you want to change concurrency, as in letting 2 or more items at the same time for a particular key, then you'd need to set the `maxCount` parameter, though I'm assuming this is not the required behaviour.

Reply:
It is not. So these configure the semaphore pool, and not the string keys. I'm going to guess I'll forget this detail at some point so a comment would be very welcome in the source code here.

Reply:
Will do.